### PR TITLE
[SPARK-9846] [Docs] User guide for Multilayer Perceptron Classifier

### PR DESCRIPTION
**Pull request by avulanov**
Added user guide for multilayer perceptron classifier:
- Simplified description of the multilayer perceptron classifier
- Example code for Scala and Java

### CODE REVIEW

File: docs/ml-ann.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+**Examples**
+
+<div class="codetabs">
+
+<div data-lang="scala" markdown="1">
+
+{% highlight scala %}
+import org.apache.spark.ml.classification.MultilayerPerceptronClassifier
+import org.apache.spark.mllib.evaluation.MulticlassMetrics
```
Comment (mengxr) [inline docs/ml-ann.md]: Use `MulticlassClassificationEvaluator` instead.

File: docs/ml-ann.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+val data = MLUtils.loadLibSVMFile(sc, "data/mllib/sample_multiclass_classification_data.txt").toDF()
+// Split the data into train and test
+val splits = data.randomSplit(Array(0.6, 0.4), seed = 1234L)
+val train = splits(0)
+val test = splits(1)
+// specify layers for the neural network: 
+// input layer of size 4 (features), two intermediate of size 5 and 4 and output of size 3 (classes)
+val layers = Array[Int](4, 5, 4, 3)
+// create the trainer and set its parameters
+val trainer = new MultilayerPerceptronClassifier().setLayers(layers).setBlockSize(128).setSeed(1234L).setMaxIter(100)
```
Comment (mengxr) [inline docs/ml-ann.md]: chop down setters

File: docs/ml-ann.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+<div data-lang="java" markdown="1">
+
+{% highlight java %}
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.SparkContext;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.ml.classification.MultilayerPerceptronClassificationModel;
+import org.apache.spark.ml.classification.MultilayerPerceptronClassifier;
+import org.apache.spark.mllib.evaluation.MulticlassMetrics;
```
Comment (mengxr) [inline docs/ml-ann.md]: Same here. Use `MulticlassClassificationEvaluator` instead.

File: docs/ml-ann.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.spark.mllib.regression.LabeledPoint;
+import org.apache.spark.mllib.util.MLUtils;
+import org.apache.spark.sql.DataFrame;
+import org.apache.spark.sql.SQLContext;
+
+public class MultilayerPerceptronClassifierTest {
+
+    public static void main( String[] args )
+    {
```
Comment (mengxr) [inline docs/ml-ann.md]: For Java example code, we don't usually put `main`.

File: docs/ml-ann.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        SparkConf conf = new SparkConf().setAppName("Multilayer perceptron classifier example").setMaster("local");
+        SparkContext sc = new SparkContext(conf);
+        SQLContext sql = new SQLContext(sc);
+        String path = "data/mllib/sample_multiclass_classification_data.txt";
+        // Load training data
```
Comment (mengxr) [inline docs/ml-ann.md]: We can start directly from here, assuming `sc` and `sqlContext` is given.

File: docs/ml-ann.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        JavaRDD<LabeledPoint> data = MLUtils.loadLibSVMFile(sc, path).toJavaRDD();
+        // Split the data into train and test
+        JavaRDD<LabeledPoint>[] splits = data.randomSplit(new double[]{0.6, 0.4}, 1234L);
+        DataFrame train = sql.createDataFrame(splits[0], LabeledPoint.class);
```
Comment (mengxr) [inline docs/ml-ann.md]: `sql` -> `sqlContext`. I would suggest creating `DataFrame` first and then call `randomSplit`.

File: docs/ml-ann.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        DataFrame test = sql.createDataFrame(splits[1], LabeledPoint.class);
+        // specify layers for the neural network:
+        // input layer of size 4 (features), two intermediate of size 5 and 4 and output of size 3 (classes)
+        int[] layers = new int[] {4, 5, 4, 3};
+        // create the trainer and set its parameters
+        MultilayerPerceptronClassifier trainer = new MultilayerPerceptronClassifier();
+        trainer.setLayers(layers).setBlockSize(128).setSeed(1234L).setMaxIter(100);
```
Comment (mengxr) [inline docs/ml-ann.md]: chain and chop down setters:

~~~java
MultilayerPerceptronClassifier trainer = new MultilayerPerceptronClassifier()
  .setLayers(layers).
  .setBlockSize(128)
  .setSeed(1234L)
  .setMaxIter(100);
~~~

**Comment by avulanov** [comment]
@mengxr Thank you for your comments! I've addressed everything except the full chop of setters for the Java trainer. It complains on `java: incompatible types: org.apache.spark.ml.classification.MultilayerPerceptronParams cannot be converted to org.apache.spark.ml.classification.MultilayerPerceptronClassifier`. I did it in two stages. Also I've left Java imports.

**Comment by avulanov** [comment]
The test fail seems unrelated to my commit.

**Comment by mengxr** [comment]
Ah, that is a bug I overlooked in review. We must define setters in the final class so the return type in Java is the self type. I will send a PR for it and ping you back.

File: docs/ml-ann.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+// Split the data into train and test
+val splits = data.randomSplit(Array(0.6, 0.4), seed = 1234L)
+val train = splits(0)
+val test = splits(1)
+// specify layers for the neural network: 
+// input layer of size 4 (features), two intermediate of size 5 and 4 and output of size 3 (classes)
+val layers = Array[Int](4, 5, 4, 3)
+// create the trainer and set its parameters
+val trainer = new MultilayerPerceptronClassifier()
+	.setLayers(layers)
```
Comment (mengxr) [inline docs/ml-ann.md]: use 2-space indentation

File: docs/ml-ann.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+// Split the data into train and test
+DataFrame[] splits = dataFrame.randomSplit(new double[]{0.6, 0.4}, 1234L);
+DataFrame train = splits[0];
+DataFrame test = splits[1];
+// specify layers for the neural network:
+// input layer of size 4 (features), two intermediate of size 5 and 4 and output of size 3 (classes)
+int[] layers = new int[] {4, 5, 4, 3};
+// create the trainer and set its parameters
+MultilayerPerceptronClassifier trainer = new MultilayerPerceptronClassifier();
+trainer.setLayers(layers)
```
Comment (mengxr) [inline docs/ml-ann.md]: #8342 was merged. So we can remove `trainier.`

**Comment by mengxr** [comment]
LGTM except some indentation issues.

**Comment by avulanov** [comment]
@mengxr Thank you. Fixed indentation and removed `trainer`. Java example works now.

**Comment by mengxr** [comment]
Merged into master and branch-1.5. Thanks!

**Pull request closed without merging**